### PR TITLE
build: upgrade to webpack 5

### PR DESCRIPTION
**Pull request by zzmp**
## Description

Upgrades the toolchain from Webpack 4 to Webpack 5. This includes some other changes required for the upgrade:
- Upgrading craco to match
- Updating the craco.config.js to deal with new requirements
- Importing JSON files using only default imports

This also includes configuring a vendors bundle, as the CRA/Webpack upgrade changed the bundle splitting behavior. Without configuring a vendors bundle, the initial JS chunk would be significantly larger and would trigger size warnings from workbox. This _should_ be more performant, as the vendors chunk _shouldn't_ change more frequently than our deploy frequency.

_Relevant docs:_ [Design doc](https://www.notion.so/uniswaplabs/Optimizing-interface-build-performance-29adef9205ae44928bac0e04ed50e382#edf3cdae66fa4e299441373bba945024), [Bug bash](https://www.notion.so/uniswaplabs/Bug-Bash-Webpack-5-Migration-8754184029394129811ecd640f1103d0)

### QA (ie manual testing)

- [x] N/A

### Automated testing

_Existing tests should continue running fine :)_

- [x] Unit test
- [x] Integration/E2E test

### CODE REVIEW

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:85]: nit: It would be great if the comment explained the difference between `debug` and `short`.

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:101]: Is this a feature (separate PR) or something to provide Webpack 4-like functionality due to change in cacheable file size as of Webpack 5?

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:102]: Do you have context why the cacheable size was changed (e.g. performance/memory) and what does that mean for us, if overwritten like this?

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:131]: Could you explain the relation between this and Webpack upgrade?

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:138]: nit: This section is missing a high-level explanation of what it does. I assume it has something to do with `@uniswap/analytics` package, but I would appreciate additional explanation.

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:154]: nit: I would appreciate if comment could be updated to mention when this can happen and why we're not interested in warnings

File: cypress/e2e/service-worker.test.ts
Comment (grabbou) [inline cypress/e2e/service-worker.test.ts:57]: Does the increased timeout means that we're having downgraded performance when it comes to loading a service worker now?

File: scripts/test-size.js
Comment (grabbou) [inline scripts/test-size.js:18]: What causes the chunk size to grow from 435 to 1790? Also, that is in KB, previously you mentioned `~7MB` in craco.config.js, so I am confused - what is the correct size here?

File: scripts/test-size.js
Comment (grabbou) [inline scripts/test-size.js:27]: Why we don't want to assert on this anymore?

File: src/state/claim/hooks.ts
Comment (grabbou) [inline src/state/claim/hooks.ts:2]: Not related to the PR, but why do we import from `build` folder? Does this package just happen to have a `build` folder or are we actually importing from `build`? Is this due to something else not working properly, such as `package.exports` not being fully supported?

**Review by grabbou — CHANGES_REQUESTED** [review]
Looking good! Great to see our infrastructure moving towards Webpack 5.x finally and dropping some of the workarounds. Left some comments, mostly nits, but I do have some questions to get more context before final approval. Looking forward.

File: craco.config.cjs
Comment (cbachmeier) [inline craco.config.cjs:128]: q: what's the difference between the resolve.plugins and the plugins we mapped through on ln 90-91?

**Review by cbachmeier — APPROVED** [review]
LGTM. Checked all pages/styles load as expected including VE global styles. Major flows such as swap, pool, purchase nft, list, work as expected on tenderly

File: craco.config.cjs
Comment (zzmp) [reply craco.config.cjs:101]: This is a required part of the migration.
Webpack 4 split the entry file into entry (very small) and main (initial assets), but Webpack 5 optimizes by serving it all as a single file. As a result, that file is much larger, and so we need to bump the limit on cacheable file size by the workbox InjectManifest plugin.

File: craco.config.cjs
Comment (zzmp) [reply craco.config.cjs:152]: No, because they are accessed two different ways. `ProvidePlugin` adds `process` to the global scope, but `path` is not on the global scope - it is still imported as a package.

File: craco.config.cjs
Comment (zzmp) [reply craco.config.cjs:131]: I think that create-react-app 5 has additional restrictions to prevent bad practices, and this is one of them - it prevents you from including files in your build which shouldn't obviously be there.

File: craco.config.cjs
Comment (zzmp) [reply craco.config.cjs:128]: These are resolver-specific plugins, whereas the other plugins have to do with the global pipeline.

File: src/state/claim/hooks.ts
Comment (zzmp) [reply src/state/claim/hooks.ts:2]: This is due to how our legacy packages are built. They export their abis in build folders. They do not use `package.exports`.

This is much older infra, so `exports` was probably just not considered at the time. I don't think we want to touch these - they are too foundational.

File: craco.config.cjs
Comment (zzmp) [reply craco.config.cjs:101]: The CRA / Webpack 5 configuration no longer splits the entry chunk, so the entry chunk (main.js) is now much larger. Instead of allowing a larger cached file, I'm going to make a vendor bundle to better mimic the Webpack 4 behavior and for better performance.

File: craco.config.cjs
Comment (zzmp) [reply craco.config.cjs:102]: See https://github.com/Uniswap/interface/pull/6459#discussion_r1180898182

File: cypress/e2e/service-worker.test.ts
Comment (zzmp) [reply cypress/e2e/service-worker.test.ts:57]: See https://github.com/Uniswap/interface/pull/6459#discussion_r1180898182

File: scripts/test-size.js
Comment (zzmp) [reply scripts/test-size.js:18]: https://github.com/Uniswap/interface/pull/6459#discussion_r1180898182

File: scripts/test-size.js
Comment (zzmp) [reply scripts/test-size.js:27]: https://github.com/Uniswap/interface/pull/6459#discussion_r1180898182

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:52]: nit: why was this not required before?

Comment (grabbou) [inline craco.config.cjs:52]: nit: what do you mean by `the defaults` in this context? Craco defaults? Webpack defaults?

File: craco.config.cjs
Comment (grabbou) [inline craco.config.cjs:84]: nit: https://uniswapteam.slack.com/archives/C04KGMR1ACW/p1682955163281759

**Review by grabbou — APPROVED** [review]

File: craco.config.cjs
Comment (zzmp) [reply craco.config.cjs]: I've updated the comment to be more explicit. CRA overrides these properties to use the eslint that ships with react-scripts, but doing so invalidates our lint cache.

Comment (zzmp) [reply craco.config.cjs]: I've updated the comment to be more explicit. CRA overrides these properties to use the eslint that ships with react-scripts.

**Review by grabbou — APPROVED** [review]

**Review by cbachmeier — APPROVED** [review]

**Review by cbachmeier — APPROVED** [review]

**Review by cbachmeier — APPROVED** [review]

**Pull request merged**
